### PR TITLE
Decoupling LRO from core-client

### PR DESCRIPTION
**Pull request by deyaaeldeen**
This PR refactors the LRO files so that all files in `src/lro` do not use anything from `core-client`. These files are being instantiated with `core-client` things in `src/coreClientLro.ts`. A similar file for `core-http` is provided, `src/coreHttpLro.ts` which restores the LRO support for core-http clients. This PR is a step toward moving everything in `src/lro` to `@azure/core-lro`. See https://github.com/Azure/azure-sdk-for-js/pull/15949 for details.

### CODE REVIEW

File: src/lro/azureAsyncPolling.ts [hunk 1]
```diff
@@ -21,15 +25,12 @@ function isAzureAsyncPollingDone(rawResponse: FullOperationResponse) {
 }
 
 export function processAzureAsyncOperationResult<TResult>(
-  restrieveResource: (path?: string) => Promise<LROResult<TResult>>,
+  lro: LRO<TResult>,
```
Comment (deyaaeldeen) [inline src/lro/azureAsyncPolling.ts:28]: passing the whole object is needed so `retrieveAzureAsyncResource` can access `this` internaly.

File: src/coreClientLRO.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+export function getSpecPath(spec: OperationSpec): string {
+  if (spec.path) {
+    return spec.path;
+  } else {
+    throw Error("Bad spec: request path is not found!");
+  }
+}
+
+export class CoreClientLRO<T> implements LRO<T> {
```
Comment (deyaaeldeen) [inline src/coreClientLRO.ts:234]: this class is the main contribution of this PR

File: src/generators/operationGenerator.ts [hunk 3]
```diff
@@ -963,16 +965,19 @@ function writeLROOperationBody(
...
   methodDeclaration.addStatements([
     sendOperationStatement,
+    `const lro = new CoreClientLRO(sendOperation,${operationParamsName},
+      ${operationSpecName},${finalStateStr})`,
     `return new LROPoller({intervalInMs: options?.updateIntervalInMs},
-      ${operationParamsName},
-      ${operationSpecName},
-      sendOperation,
-      ${finalStateStr}
+      lro
```
Comment (deyaaeldeen) [inline src/generators/operationGenerator.ts:980]: this is where `CoreClientLRO` is instantiated.

File: src/lro/models.ts [hunk 4]
```diff
@@ -25,20 +43,89 @@ export interface LROConfig {
...
+export type RetrieveAzureAsyncResourceOperation<T> = PollingOperation<T>;
+export type GetLROState<T> = (
+  rawResponse: RawResponse,
+  flatResponse: T
+) => LROState<T>;
+
+/**
+ * Description of a long running operation
+ */
+export interface LRO<T> {
```
Comment (deyaaeldeen) [inline src/lro/models.ts:110]: this is the abstract interface to be instantiated by any client

File: src/transforms/operationTransforms.ts [hunk 5]
```diff
@@ -133,7 +133,13 @@ export function transformOperationSpec(
...
 }
 
-export function extractHttpDetails({ path, method }: OperationRequestDetails) {
+export function extractHttpDetails({
+  path,
+  method
+}: OperationRequestDetails): {
+  path: string;
+  httpMethod: HttpMethods;
+} {
```
Comment (deyaaeldeen) [inline src/transforms/operationTransforms.ts:142]: for some reason the compiler did not like leaving the return type implicit so I had to add it.

**Comment by sarangan12** [comment]
The changes look fine. But, in our last conversation, Jose was telling me that the plan was to move LROEngine to core-lro. This PR does not seem to do that. Or is it some intermediate step? Can you clarify?

**Comment by deyaaeldeen** [comment]
@sarangan12 sorry about the confusion! Yes it is an intermediate step. I updated the PR description.

**Review by sarangan12 — APPROVED** [review]

**Pull request merged**